### DISCUSSION TITLE
Frequency list generator

### DISCUSSION THREAD

**Discussion by mortii** [mortii/anki-morphs#36] [Ideas]
[I mentioned in issue #17](https://github.com/mortii/anki-morphs/issues/17#issuecomment-1775012244):

> I found really bad problems with the morphman implementation of frequency.txt... It does not differentiate between between inflected morphs and base morphs, so it will almost certainly misinterpret morphs and give false-positives.

The generator should be able to output the morphs in a base (norm) + inflected format. I'm not sure about the best file format for this, probably json?

There should probably also be an option to produce a more human readable file, e.g. morphman style,  where the output is only the inflected version of the morph.

**Comment by Vilhelm-Ian** [comment]
I love json I am all up for it. Csv is a different option since we would be only storing two variables per line. It's not as descriptive as json but simpler

**Reply by mortii** [reply]
Ah, I hadn't even thought of csv, that is a good alternative. I don't know which format has a more efficient read-write performance in python. Maybe we could support both if it's not too much trouble?

**Reply by mortii** [reply]
actually, if we use csv then that also covers the 'human readable' option so that would kill two birds in one stone no?

**Reply by Vilhelm-Ian** [reply]
Yes for the human readable format.

My guess is that csv will be faster. Primagen has a video why json is slow https://www.youtube.com/watch?v=MuCK81q1edU

**Reply by mortii** [reply]
[This comment](https://stackoverflow.com/questions/26156646/which-one-is-best-csv-or-json-in-order-to-import-big-data-php/26161739#26161739) also agrees. Using csv is definitely better than json for our usecase.

**Comment by mortii** [comment]
How do we want the csv file to be formatted?

I think it should be top-down with the most frequent morphs at the top, and some column header/title at the first line, e.g:
``` text
Morph-base,Morph-inflection
a,aa
b,bb
```

**Reply by Vilhelm-Ian** [reply]
that's how I imagined it

**Reply by aleksejrs** [reply]
The morphemizer could keep commas, maybe use tabs instead.

**Reply by aleksejrs** [reply]
Also, I like to see the number of occurrences in the file and overall (though that probably slows grepping down). It would be better if my symlinks didn't modify the numbers though.

**Reply by Vilhelm-Ian** [reply]
Why would you like to see the number of occurencecs

**Reply by mortii** [reply]
> The morphemizer could keep commas, maybe use tabs instead.

Could they? I don't think that should happen anymore after the #28 fix.

**Reply by mortii** [reply]
> Also, I like to see the number of occurrences in the file

I think the frequency file should be as minimal as possible for performance reasons, it is going to be read by the addon many, many times.

Things like the number of occurrences should be restricted to the [Readability report (#37) ](https://github.com/mortii/anki-morphs/discussions/37 ) imo.

**Reply by aleksejrs** [reply]
> Could they? I don't think that should happen anymore after the #28 fix.

I considered including punctuation when getting my morphemizer to use two-word morphs. It seems that I didn't do that, but it depends on the morphemizer.

**Reply by aleksejrs** [reply]
> Why would you like to see the number of occurencecs

Sometimes I use frequency.txt to decide what words to add cards with.

**Comment by mortii** [comment]
@Vilhelm-Ian I made a rough 'frequency file generator'-skeleton in the [latest commit](https://github.com/mortii/anki-morphs/commit/5f4336bc883c889048caeb49261afc00734b1866). Do you want to work on it?

**Reply by mortii** [reply]
I forgot to add the minimum occurrence option, I'll add it tomorrow.

**Reply by Vilhelm-Ian** [reply]
I tried to work on it but got this error
```Traceback (most recent call last):
  File "/home/user/programiranje/forks/anki-morphs/venv/lib/python3.11/site-packages/aqt/progress.py", line 118, in handler
    func()
  File "/home/user/programiranje/forks/anki-morphs/venv/lib/python3.11/site-packages/aqt/main.py", line 218, in on_window_init
    fn()
  File "/home/user/programiranje/forks/anki-morphs/venv/lib/python3.11/site-packages/aqt/main.py", line 317, in setupProfile
    self.loadProfile()
  File "/home/user/programiranje/forks/anki-morphs/venv/lib/python3.11/site-packages/aqt/main.py", line 496, in loadProfile
    gui_hooks.profile_did_open()
  File "/home/user/programiranje/forks/anki-morphs/venv/lib/python3.11/site-packages/_aqt/hooks.py", line 4101, in __call__
    hook()
  File "/home/user/.local/share/Anki2/addons21/ankimorphs/__init__.py", line 191, in init_db
    read_config_filters: list[AnkiMorphsConfigFilter] = get_read_enabled_filters()
                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/Anki2/addons21/ankimorphs/config.py", line 147, in get_read_enabled_filters
    config_filters = _get_filters_config()
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/Anki2/addons21/ankimorphs/config.py", line 191, in _get_filters_config
    filters.append(AnkiMorphsConfigFilter(_filter))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/Anki2/addons21/ankimorphs/config.py", line 29, in __init__
    self.unknowns_field_: str = _get_filter_str(_filter, "unknowns_field")
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/.local/share/Anki2/addons21/ankimorphs/config.py", line 241, in _get_filter_str
    filter_item = _filter[key]
                  ~~~~~~~^^^^^
KeyError: 'unknowns_field'

```

**Reply by Vilhelm-Ian** [reply]
Why should I setup a virtual enviorment(I have almost no experience with python)

**Reply by mortii** [reply]
> I tried to work on it but got this error
{...}
KeyError: 'unknowns_field'

Oh right, I renamed the 'Focus Morph' extra field to 'Unknowns', I think it is more descriptive. The field is optional in Ankimorphs, not a requirement like it was in morphman. If you save settings again this error should go away I think.

**Reply by mortii** [reply]
> Why should I setup a virtual enviorment(I have almost no experience with python)

- less package clutter on your pc (global environment)
- makes sure you are using the right version of the packages
- makes sure the packages are consistent for all developers
- reduce the risk of package conflicts

This last point is particularly useful.

**Reply by mortii** [reply]
@Vilhelm-Ian On an unrelated note: There is no deadline for this project, so if you were to only work on this for 15 minutes every other month or something, that would be completely fine. Don't feel any pressure to finish as soon as possible. Stay healthy :pray:

**Reply by Vilhelm-Ian** [reply]
@mortii  I got it working with deleting the meta.json

**Reply by Vilhelm-Ian** [reply]
I can't work in moderation. Only excess :(

**Comment by mortii** [comment]
It occurred to me that we should probably also include the relevant 'parse' options that are in settings.

**Comment by mortii** [comment]
I just remembered about the ```Allow for custom morph prioritization for specified languages (e.g. jp-morph-priority.txt)``` item in the roadmap. I think I have a good solution for this: have a frequency list drop down menu in note filters.

Say you have two note filters, one for german cards and one for japanese cards it would look something like this:
```
Note Type                   Tags            ...          Prioritization                 
1. german_sub2srs           ""              ...          de_frequency_list.csv  
2. japanese_sub2srs         ""              ...          jp_frequency_list.csv  
```

**Comment by Vilhelm-Ian** [comment]
@mortii This is the route I want to take for selecting files. If the user selects multiple files in first lineEdit to display only file names and in the class to store the path of the files and when the user presses generate frequency list to concate the path stored in the class with the names. And if the user instead decides to paste a path in the input field to check if it's a valid path and if it's to use that.

here is some pseudo-code
```
def generate_freq() {
    if isFile(lineEdit.text) {
       files = lineEdit.text.split(" ")
        for(file in this.file) {
           if isFile(this.path+file) {
            this.text +=  # The text located in this.path+file
           }
        }
    }
}
```

Because if the user selects multiple files the input field will get too clutered
![image](https://github.com/mortii/anki-morphs/assets/81630959/8c0326d1-6108-4c95-b926-7f7527144272)

I like this route better than having to select a directory because it allows the user the freedom to chose which files.

**Reply by mortii** [reply]
That is such a good idea, I completely agree!

**Reply by mortii** [reply]
We need to make sure selecting multiple files is cross-platform compatible. Hopefully there are a million guides on how to do that.

**Reply by Vilhelm-Ian** [reply]
I think it should be since Qt it self is cross platform.  I mean it's file picker.

**Reply by aleksejrs** [reply]
Will it still be possible to select an input directory?

**Reply by mortii** [reply]
@aleksejrs with the current implementation I have been using Ctrl+A to select everything in the dir. Do you think that is a bad approach?

**Reply by aleksejrs** [reply]
@mortii I have a directory tree with thousands of files, including directories with github wikis and book chapters (ocrmypdf output or manually split to prioritize what I am reading).

**Reply by mortii** [reply]
@aleksejrs this approach allows for a more granular selection as well as selecting everything, so it is better in every way no?

**Reply by mortii** [reply]
@aleksejrs This method does not select files recursively, no directories can be selected, only current dir files. This is a good thing in my opinion.

**Reply by aleksejrs** [reply]
@mortii How do I select the files in different subdirectories?

**Reply by mortii** [reply]
@aleksejrs I'm open to different approaches, if you think recursive is a good feature then that could probably be included.

**Reply by aleksejrs** [reply]
@mortii Recursive is definitely a good feature. MorphMan uses only two lists, so there are only two sets of files max (unless you are interested in Readability Analyzer). For prioritization, I concatenated the files in some directories, but that's inconvenient and I think MorphMan already supports per-directory calculation in **some** way, but since I am using an old version, I don't know how exactly.

**Reply by mortii** [reply]
@aleksejrs Ok, i'll try to find a good approach that supports recursion. I'm envisioning a tree view selector, maybe that could work.

**Comment by mortii** [comment]
I updated the frequency file dialog to this:
![Screenshot from 2023-11-14 15-11-27](https://github.com/mortii/anki-morphs/assets/15674619/26f405a6-d944-48e0-917d-1eb125e5c220)

The output file is automatically set to be `profilefolder/frequency-files/frequency.csv`. This is the directory where I want to extract file names from in my proposed new note-filters option.

**Comment by mortii** [comment]
### Use-case:
A user has a folder 'subs' and that folder has multiple sub-directories, one for each tv-show, each of which contains multiple srt files.
If the user selects the 'sub' directory then all the sub-directories should be recursively traversed and all the srt files gathered.

### QFileDialog
QFileDialog can either open selected file(s) or a selected directory, it cannot do both. If we want to support the use-case above then we would have to go for selecting directories only. Since volume (usually) makes the frequency files better I think it is more important to support this greedy approach rather than selecting individual files.

**Reply by Vilhelm-Ian** [reply]
I think the only way to satisfy both file selection and dir selection is to have  two separate buttons. One select files, other select directories. The problem with this is it adds complexity for the user

**Reply by mortii** [reply]
@Vilhelm-Ian I agree. Let's not bother with individual files for now. We can revisit this later If there turns out to be a desperate need for it by a substantial amount of users (doubt it).

**Reply by mortii** [reply]
@aleksejrs I have a working version of this that gathers txt and srt files. Are there other formats you want added?

**Reply by mortii** [reply]
Now that I think about it, there should probably be an option for which file formats the user wants gathered.

**Reply by aleksejrs** [reply]
> @aleksejrs I have a working version of this that gathers txt and srt files. Are there other formats you want added?

* .vtt (WEBVTT) — subtitles from YouTube, very useful and supported by [movies2anki](https://github.com/kelciour/movies2anki).
* .ass (Script generated by Aegisub) — a more detailed subtitle format, also supported by movies2anki. I have only a few files in it that can be converted.
* .md (Markdown) — the extension should be supported.
----
* HTML — it would be much better if it was supported directly. HTML files are easy to download, but conversion requires special settings to avoid_strings_like_this.
* .epub — it's probably almost as easy as HTML.
 ----
* .ssa (Sub Station Alpha v4 script) — uselessly rare on my computer, but seems like a version of .ass.

**Reply by mortii** [reply]
- .vtt can be added without any conversion so that is fine
- anki has comes with a strip_html utils function so if that works well then the html related formats could be added
      - .md (could contain html)
      - .html
      - .epub

.ass and .ssa needs custom stripping, if you want those added then PRs are open.

**Reply by mortii** [reply]
ok, epub is non-trivial (see https://github.com/aerkalov/ebooklib/) so supporting that is out of scope.

**Reply by mortii** [reply]
bad new for the html options:
- for the html strip to work the entire file needs to be loaded to memory first, which is really bad.
- the stripper is obviously not made for our use-case and therefore it creates _alot_ of concatenated words, so the resulting text is of poor quality.

Because of these two problems I deem html related files not worth supporting, unfortunately.

**Reply by aleksejrs** [reply]
Then md should be supported as txt.

**Reply by mortii** [reply]
@aleksejrs sure, I'll add it.

**Reply by mortii** [reply]
@aleksejrs done.

**Comment by mortii** [comment]
![frequency-file-generator](https://github.com/mortii/anki-morphs/assets/15674619/71db7588-94e9-416c-bf8e-90f06d3a2ba7)

This is the current version. I'm satisfied with it so I'm going to close this thread.

**Reply by mortii** [reply]
I updated the docs: https://mortii.github.io/anki-morphs/user_guide/usage/frequency-file-generator.html#select-input
hopefully it makes sense.

**Reply by aleksejrs** [reply]
It says vff instead of vtt.

**Reply by mortii** [reply]
@aleksejrs oops, thank you!
